# Eine kurze Einführung über Datentypen in Python

<img src="https://i.imgur.com/FqtUbJd.png" width=800 />


Dieses Tutorial behandelt die folgenden Themen:

- Speichern von Informationen mit Variablen
- Primitive Datentypen in Python: Integer, Float, Boolean, None und String
- Eingebaute Datenstrukturen in Python: Liste, Tupel und Wörterbuch
- Von eingebauten Datentypen unterstützte Methoden und Operatoren

## Speichern von Informationen mit Hilfe von Variablen

Computer sind für zwei Zwecke nützlich: zum Speichern von Informationen (auch Daten genannt) und zum Ausführen von Operationen mit gespeicherten Daten. Wenn Sie mit einer Programmiersprache wie Python arbeiten, werden Daten in Variablen gespeichert. Man kann sich Variablen als Behälter zum Speichern von Daten vorstellen. Die Daten, die in einer Variablen gespeichert sind, werden als ihr Wert bezeichnet. Das Erstellen von Variablen in Python ist ziemlich einfach, wie wir bereits im [vorherigen Lernprogramm](https://jovian.ml/aakashns/first-steps-with-python/v/4#C15) gesehen haben.



In [7]:
meine_lieblingsfarbe = "grün"

In [2]:
meine_lieblingsfarbe = "grün"

In [8]:
meine_lieblingsfarbe

'grün'

Eine Variable wird mit einer Zuweisungsanweisung erstellt. Sie beginnt mit dem Variablennamen, gefolgt von dem Zuweisungsoperator `=` und dem Wert, der in der Variablen gespeichert werden soll.  Beachten Sie, dass der Zuweisungsoperator `=` sich von dem Gleichheitsvergleichsoperator `==` unterscheidet.

Sie können auch mehreren Variablen in einer einzigen Anweisung Werte zuweisen, indem Sie die Variablennamen und -werte mit Kommas trennen.

In [3]:
color1, color2, color3 = "rot", "grün", "blau"

In [4]:
color1

'rot'

In [5]:
color2

'grün'

In [6]:
color3

'blau'

Sie können mehreren Variablen denselben Wert zuweisen, indem Sie mehrere Zuweisungsoperationen innerhalb einer einzigen Anweisung verketten.

In [9]:
color4 = color5 = color6 = "magenta"

In [10]:
color4

'magenta'

In [11]:
color5

'magenta'

In [12]:
color6

'magenta'

Sie können den in einer Variablen gespeicherten Wert ändern, indem Sie ihr mit einer anderen Zuweisungsanweisung einen neuen Wert zuweisen. Seien Sie bei der Neuzuweisung von Variablen vorsichtig: Wenn Sie der Variablen einen neuen Wert zuweisen, geht der alte Wert verloren und ist nicht mehr zugänglich.

In [13]:
lieblingsfarbe = "grün"

In [14]:
lieblingsfarbe

'grün'

Wenn Sie eine Variable neu zuweisen, können Sie auch den vorherigen Wert der Variable verwenden, um den neuen Wert zu berechnen.

In [15]:
zaehler = 10

In [16]:
zaehler = zaehler + 1

In [17]:
zaehler

11

Das Muster `var = var op irgendwas` (wobei `op` ein arithmetischer Operator wie `+`, `-`, `*`, `/` ist) ist sehr gebräuchlich, daher bietet Python eine *Kurzschrift* dafür.

In [ ]:
zaehler = 10

In [ ]:
# das gleiche wie `zaehler = zaehler + 4`
zaehler += 4

In [ ]:
zaehler

### gültige Variablennamen

Variablennamen können kurz (`a`, `x`, `y`, usw.) oder beschreibend sein ( `my_favorite_color`, `profit_margin`, `the_3_musketeers`, usw.). Bei der Benennung von Python-Variablen müssen Sie jedoch die folgenden Regeln beachten:

* Der Name einer Variablen muss mit einem Buchstaben oder dem Unterstrich `_` beginnen. Er darf nicht mit einer Zahl beginnen.
* Ein Variablenname darf nur Klein- oder Großbuchstaben, Ziffern oder Unterstriche enthalten (`a`-`z`, `A`-`Z`, `0`-`9` und `_`).
* Bei Variablennamen wird zwischen Groß- und Kleinschreibung unterschieden, d.h. `a_variable`, `A_Variable` und `A_VARIABLE` sind alle unterschiedliche Variablen.

Hier sind einige gültige Variablennamen:

In [ ]:
a_variable = 23
is_today_Saturday = False
my_favorite_car = "Delorean"
the_3_musketeers = ["Athos", "Porthos", "Aramis"] 

Lassen Sie uns versuchen, einige Variablen mit ungültigen Namen zu erstellen. Python gibt einen Syntaxfehler aus, wenn der Name Ihrer Variablen ungültig ist.

> **Syntax**: Die Syntax einer Programmiersprache bezieht sich auf die Regeln, die die Struktur einer gültigen Anweisung oder *Anweisung* bestimmen. Wenn eine Anweisung diesen Regeln nicht folgt, hält Python die Ausführung an und informiert Sie über einen *Syntaxfehler*. Sie können sich die Syntax als die Regeln der Grammatik einer Programmiersprache vorstellen.

In [ ]:
a variable = 23

In [ ]:
is_today_$aturday = False

In [ ]:
my-favorite-car = "Delorean"

In [ ]:
3_musketeers = ["Athos", "Porthos", "Aramis"]

## Eingebaute Datentypen in Python

Alle Daten oder Informationen, die in einer Python-Variablen gespeichert sind, haben einen *Typ*. Sie können den Typ der in einer Variablen gespeicherten Daten mit der Funktion `type` anzeigen.

In [ ]:
a_variable

In [ ]:
type(a_variable)

In [ ]:
is_today_Saturday

In [ ]:
type(is_today_Saturday)

In [ ]:
my_favorite_car

In [ ]:
type(my_favorite_car)

In [ ]:
the_3_musketeers

In [ ]:
type(the_3_musketeers)

Python hat mehrere eingebaute Datentypen, um verschiedene Arten von Informationen in Variablen zu speichern. Im Folgenden sind einige häufig verwendete Datentypen aufgeführt:

1. integer (Ganzzahl)
2. float (Fließkomma)
3. boolean (Boolescher Wert)
4. None (Kein Wert)
5. String (Zeichenfolge)
6. Liste
7. Tupel
8. Dictionary (Wörterbuch)

Integer, Float, Boolean, None und String sind *primitive Datentypen*, weil sie einen einzigen Wert darstellen. Andere Datentypen wie Liste, Tupel und Wörterbuch werden oft als *Datenstrukturen* oder *Container* bezeichnet, weil sie mehrere Daten zusammenhalten. Container betrachten wir hier nur kurz. Wir werden später diese genauer untersuchen.

### Integer

Ganzzahlen stehen für positive oder negative ganze Zahlen, von negativ unendlich bis unendlich. Beachten Sie, dass Ganzzahlen keine Dezimalpunkte enthalten sollten. Ganzzahlen haben den Typ `int`.

In [ ]:
current_year = 2020

In [ ]:
current_year

In [ ]:
type(current_year)

Im Gegensatz zu einigen anderen Programmiersprachen können ganze Zahlen in Python beliebig groß (oder klein) sein. Es gibt keinen niedrigsten oder höchsten Wert für Ganzzahlen, und es gibt nur einen `int`-Typ (im Gegensatz zu `short`, `int`, `long`, `long long`, `unsigned int`, etc. in C/C++/Java).

In [4]:
a_large_negative_number = -23374038374832934334234317348343

In [5]:
a_large_negative_number

-23374038374832934334234317348343

In [6]:
type(a_large_negative_number)

int

### Fließkommazahl

Fließkommazahlen (oder Gleitkommazahlen) sind Zahlen mit einem Dezimalpunkt. Es gibt keine Grenzen für den Wert oder die Anzahl der Stellen vor oder nach dem Dezimalpunkt. Fließkommazahlen haben den Typ `Float`.

In [7]:
pi = 3.141592653589793238

In [8]:
pi

3.141592653589793

In [9]:
type(pi)

float

In [11]:
k = 1
for i in range(8):
    print(k, type(k))
    k+=0.5

1 <class 'int'>
1.5 <class 'float'>
2.0 <class 'float'>
2.5 <class 'float'>
3.0 <class 'float'>
3.5 <class 'float'>
4.0 <class 'float'>
4.5 <class 'float'>


Beachten Sie, dass eine ganze Zahl als Fließkommazahl behandelt wird, wenn sie mit einem Dezimalpunkt geschrieben wird, auch wenn der Dezimalteil der Zahl Null ist.

In [ ]:
a_number = 3.0

In [ ]:
a_number

In [ ]:
type(a_number)

In [ ]:
another_number = 4.

In [ ]:
another_number

In [ ]:
type(another_number)

Fließkommazahlen können auch in wissenschaftlicher Notation geschrieben werden, wobei ein "e" die Potenz von 10 angibt.

In [ ]:
one_hundredth = 1e-2

In [ ]:
one_hundredth

In [ ]:
type(one_hundredth)

In [ ]:
avogadro_number = 6.02214076e23

In [ ]:
avogadro_number

In [ ]:
type(avogadro_number)

Mit den Funktionen `float` und `int` können Sie Fließkommazahlen in Ganzzahlen umwandeln und umgekehrt. Der Vorgang der Umwandlung eines Wertes in einen anderen Typ wird als *Casting* bezeichnet.

In [ ]:
float(current_year)

In [ ]:
float(a_large_negative_number)

In [ ]:
int(pi)

In [ ]:
int(avogadro_number)

Bei der Durchführung arithmetischer Operationen werden Ganzzahlen automatisch in Fließkommazahlen umgewandelt, wenn einer der Operanden eine Fließkommazahl ist. Auch der Divisionsoperator `/` gibt immer eine "Fließkommazahl" zurück, selbst wenn beide Operanden Ganzzahlen sind. Verwenden Sie den `//`-Operator, wenn das Ergebnis der Division eine `Inte` sein soll.

In [ ]:
type(45 * 3.0)

In [ ]:
type(45 * 3)

In [ ]:
type(10/3)

In [ ]:
type(10/2)

In [ ]:
type(10//2)

### Boolesche Variablen

Boolesche Werte stellen einen von 2 Werten dar: `True` und `False`. Boolesche Werte haben den Typ `bool`.

In [ ]:
ist_heute_Sonntag = True

In [ ]:
ist_heute_Sonntag

In [ ]:
type(ist_heute_Sonntag)

Boolesche Werte sind im Allgemeinen das Ergebnis einer Vergleichsoperation, z. B. `==`, `>=` usw.

In [ ]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10

In [ ]:
is_ice_bag_expensive

In [ ]:
type(is_ice_bag_expensive)

Boolesche Werte werden automatisch in `int` umgewandelt, wenn sie in arithmetischen Operationen verwendet werden. Wahr" wird in "1" und "Falsch" in "0" umgewandelt.

In [ ]:
5 + False

In [ ]:
3. + True

Jeder Wert in Python kann mit der Funktion `bool` in einen boolschen Wert umgewandelt werden. 

Nur die folgenden Werte werden als `False` ausgewertet (sie werden oft *falsy* Werte genannt):

1. Der Wert `False` selbst
2. Die Ganzzahl `0`
3. Die Fließkommazahl `0.0`
4. Der leere Wert `None`
5. Der leere Text `""`
6. Die leere Liste `[]`
7. Das leere Tupel `()`
8. Das leere Wörterbuch `{}`
9. Die leere Menge `set()`
10. Der leere Bereich `Bereich(0)`

Alles andere wird als `Wahr` ausgewertet (ein Wert, der als `Wahr` ausgewertet wird, wird oft ein *wahrer* Wert genannt).

In [ ]:
bool(False)

In [ ]:
bool(0)

In [ ]:
bool(0.0)

In [ ]:
bool(None)

In [ ]:
bool("")

In [ ]:
bool([])

In [ ]:
bool(())

In [ ]:
bool({})

In [ ]:
bool(set())

In [ ]:
bool(range(0))

In [ ]:
bool(True), bool(1), bool(2.0), bool("hello"), bool([1,2]), bool((2,3)), bool(range(10))

### None

Der Typ None umfasst einen einzigen Wert `None`, der verwendet wird, um das Fehlen eines Wertes anzuzeigen. `None` hat den Typ `NoneType`. Er wird häufig verwendet, um eine Variable zu deklarieren, deren Wert später zugewiesen werden kann.

In [1]:
nothing = None

In [2]:
type(nothing)

NoneType

### Zeichenketten (Strings)

Ein String wird in Python zur Darstellung von Text (*eine Zeichenkette*) verwendet. Strings müssen in Anführungszeichen gesetzt werden (entweder das einfache Anführungszeichen `'` oder das doppelte Anführungszeichen `"`). Strings haben den Typ `string`.

In [ ]:
today = "Saturday"

In [ ]:
today

In [ ]:
type(today)

Sie können einfache Anführungszeichen innerhalb einer mit doppelten Anführungszeichen geschriebenen Zeichenfolge verwenden und umgekehrt.

In [ ]:
my_favorite_movie = "One Flew over the Cuckoo's Nest" 

In [ ]:
my_favorite_movie

In [ ]:
my_favorite_pun = 'Thanks for explaining the word "many" to me, it means a lot.'

In [ ]:
my_favorite_pun

Um ein doppeltes Anführungszeichen innerhalb einer mit doppelten Anführungszeichen geschriebenen Zeichenfolge zu verwenden, *entfernen* Sie die inneren Anführungszeichen, indem Sie ihnen das Zeichen `\` voranstellen.

In [ ]:
another_pun = "The first time I got a universal remote control, I thought to myself \"This changes everything\"."

In [ ]:
another_pun

Mit einfachen oder doppelten Anführungszeichen erstellte Zeichenketten müssen in derselben Zeile beginnen und enden. Um mehrzeilige Zeichenfolgen zu erstellen, verwenden Sie drei einfache Anführungszeichen `'''` oder drei doppelte Anführungszeichen `"""`, um die Zeichenfolge zu beginnen und zu beenden. Zeilenumbrüche werden mit dem Zeilenumbruchzeichen `\n` dargestellt.

In [ ]:
yet_another_pun = '''Son: "Dad, can you tell me what a solar eclipse is?" 
Dad: "No sun."'''

In [ ]:
yet_another_pun

Mehrzeilige Zeichenketten werden am besten mit der Funktion `print` angezeigt.

In [ ]:
print(yet_another_pun)

In [ ]:
a_music_pun = """
Two windmills are standing in a field and one asks the other, 
"What kind of music do you like?"  

The other says, 
"I'm a big metal fan."
"""

In [ ]:
print(a_music_pun)

Sie können die Länge einer Zeichenkette mit der Funktion `len` überprüfen.

In [ ]:
len(my_favorite_movie)

Beachten Sie, dass Sonderzeichen wie `\n` und maskierte Zeichen wie `\"` als ein einziges Zeichen zählen, auch wenn sie als zwei Zeichen geschrieben und manchmal auch gedruckt werden.

In [ ]:
multiline_string = """a
b"""
multiline_string

In [ ]:
len(multiline_string)

Zwei oder mehr Zeichenketten können mit dem Operator `+` verbunden oder *verkettet* werden. Seien Sie vorsichtig beim Verketten von Zeichenketten, manchmal müssen Sie ein Leerzeichen `" "` zwischen den Wörtern einfügen.

In [ ]:
full_name = "Derek O'Brien"

In [ ]:
greeting = "Hello"

In [ ]:
greeting + full_name

In [ ]:
greeting + " " + full_name + "!" # additional space

## Datenstrukturen

Wir haben uns mit den primitiven Datentypen in Python beschäftigt. Die nicht-primitiven Datenstrukturen sehen wir uns später an. Hier nur das wichtigste:

- Eine Liste ist eine Datenstruktur welche viele verschiedene Daten in einer frei wählbaren Reihenfolge zusammenfasst.
- Ein Tupel ist ähnlich wie eine Liste nur kann man die Werte nach dem erstellen nicht mehr ändern.
- Wörterbücher (Dictionarys) ist eine Sammlung von Objekten welche über einen Schlüssel identifiziert wird. Ein Schlüssel kann dabei eine Zahl oder ein Text sein.

Genaueres lernen wir in den entsprechenden Abschnitten:
- [Listen und Tupel](09_Listen.ipynb)
- [Wörterbücher](10_Dictionarys.ipynb)


### Listen

Wie der Name schon sagt eine Liste von Werten. Eine Liste wird mit eckigen Klammern geschrieben. Die einzelnen Elemente können über eine interne Nummerierung referenziert werden.

~~~python
a = [3,"hi", None, 2.71828]
b = a[1]  # b wird der Wert "hi" zugewiesen
~~~

### Tupel

Ein Tupel ist eine Liste bei welcher die einzelnen Elemente nicht mehr verändert werden können. Ein Tupel wird mit runden Klammen geschrieben.

~~~python
t = (3,2,"hi there")
~~~

[Was ist mutable/immutable?](https://www.data-science-architect.de/mutable-und-immutable-objects/)


### Wörterbuch/Dictionary

Ein Wörterbuch sammelt mehrere Objekte in einer Einheit zusammen. Im Gegensatz zu einer Liste kann man die Informationen aber über einen Schlüsselbegriff identifizieren. Wörterbücher schreibt man mit geschweiften Klammern.

~~~python
d = {"hi": 4, 2: None, 3.14: "Pi"}
e = d[3.14]  # e nimmt den Wert "pi" an.
~~~